### PR TITLE
Fix fast-dispatch hang on pinned interleaved DRAM writes (unaddressable IOVA)

### LINKED ISSUE #49691: Fast-dispatch hang on pinned interleaved DRAM writes: IOMMU returns a NOC-unreachable IOVA
## Summary
Writing an interleaved DRAM buffer via fast dispatch on the **pinned-memory write path** (used by `ttnn.from_torch(..., device=..., DRAM)` and any pinned `EnqueueWrite`) can wedge the device. First seen as a >~512 MiB size effect, but the real cause is an **unaddressable IOVA**, not size.

## Root cause (updated — fully diagnosed)
Diagnosed via watcher + tt-triage (over NOC1, since NOC0 was hung), then by instrumenting the host to log the buffer's device IOVA.

1. The pinned write path emits a **single unchunked** `CQ_PREFETCH_CMD_RELAY_LINEAR` for the whole buffer (`write_interleaved_buffer_to_device` does not iterate for pinned transfers), reading the source straight from the buffer's device **IOVA**.
2. On an IOMMU host, `PCIDevice::map_for_dma` (`TENSTORRENT_IOCTL_PIN_PAGES`) returns a kernel-assigned IOVA. Small buffers land in a **low reserved IOVA region** (base `0x20000000`); once a buffer no longer fits, the IOMMU hands back a **top-down 64-bit IOVA** near 2^64.
3. The Blackhole NOC can only express a **36-bit local address** for PCIe/host reads (`NOC_ADDR_LOCAL_BITS = 36`; `NOC_LOCAL_ADDR(addr) = addr & 0x1000000FFFFFFFFF`). A top-down IOVA has bits far above bit 36, so `cq_prefetch::process_relay_linear_cmd` reads a wrong/unmapped address, `noc_async_read_barrier_with_trid` never returns (`noc_reads_num_issued > NIU_MST_RD_RESP_RECEIVED`), and the device wedges.

### Measured IOVAs (fresh BH P100)
| buffer size | IOVA base | NOC-reachable? | result |
|---|---|---|---|
| 448–500 MiB | `0x2000_0040` | yes | OK |
| 508–512 MiB | `0xffffffff_e0000040` (≈ 2^64 − 512 MiB) | no | **hang** |

The `map_for_dma` return is `__u64` and passes to the NOC read verbatim — the top-down value is what the kernel/IOMMU assigns, not a UMD/tt-metal truncation.

## "Exact limit"
There is **no fixed size limit**. The constraint is IOVA addressability: the buffer's IOVA range must lie within the NOC-addressable PCIe window (`hal.get_pcie_addr_lower_bound()..get_pcie_addr_upper_bound()`; `< 2^36` local on BH). The ~512 MiB number is just where *this* host's IOMMU allocator switches from the low region to top-down IOVAs — it depends on the allocator/kernel/occupancy and is not portable. A smaller buffer can also hang if the low region is already occupied.

## Fix in this PR (#49695) — defensive guard
Only take the pinned fast-path when the buffer's IOVA range is NOC-addressable; otherwise fall back to the non-pinned chunked hugepage path (whose command-queue IOVA is always addressable). Guard is on addressability, not size. Regression test: `distributed_unit_tests` → `MeshBufferTestSuite.EnqueueWritePinnedMemoryLargeInterleavedBuffer` (hangs without the guard, passes with it).

## Root fix (separate)
Constrain the device DMA mask so the IOMMU never allocates unreachable IOVAs — tracked in #49696. With that, large pinned writes work directly and this guard becomes a redundant safety net.

## Affected tests
`tests/ttnn/nightly/.../deepseek_prefill/test_deepseek_moe_post_combine_reduce.py::test_multi_chunk_{structured,random}` (num_tokens > ~4096, models with emb_dim >= ~6144). Masked on `main` because the nightly `-x` session aborts at an earlier flake first.

## Repro
```python
import torch, ttnn
dev = ttnn.open_device(device_id=0)
t = torch.zeros(1, 6400, 8, 7168, dtype=torch.bfloat16)  # ~734 MiB
tt = ttnn.from_torch(t, device=dev, layout=ttnn.ROW_MAJOR_LAYOUT, memory_config=ttnn.DRAM_MEMORY_CONFIG)
ttnn.synchronize_device(dev)  # hangs
```
Recovery: `tt-smi -r 0`.

### PR DESCRIPTION
**Pull request by pavlejosipovic**
## Problem
Writing an interleaved DRAM buffer via fast dispatch on the pinned-memory write path (`ttnn.from_torch(..., DRAM)` and any pinned `EnqueueWrite`) can wedge the device.

## Root cause (fully diagnosed — not a size limit)
The pinned write path emits a single unchunked `CQ_PREFETCH_CMD_RELAY_LINEAR` that reads the source from the buffer's device **IOVA**. On an IOMMU host, `map_for_dma` returns a kernel-assigned IOVA; once a buffer no longer fits the low reserved IOVA region, the IOMMU hands back a **top-down 64-bit IOVA** (near 2^64). The Blackhole NOC only expresses a **36-bit local address** for PCIe reads (`NOC_ADDR_LOCAL_BITS = 36`), so `cq_prefetch::process_relay_linear_cmd` reads a wrong/unmapped address, `noc_async_read_barrier_with_trid` never returns, and the device hangs.

Measured on a fresh BH P100:

| buffer size | IOVA base | NOC-reachable? | result |
|---|---|---|---|
| 448–500 MiB | `0x2000_0040` | yes | OK |
| 508–512 MiB | `0xffffffff_e0000040` | no | hang |

The ~512 MiB threshold is **not** a hardware limit — it is where this host's IOMMU allocator switches to top-down IOVAs, and it depends on the allocator/kernel/occupancy (a smaller buffer can also hang if the low region is occupied). Diagnosed via watcher + tt-triage (NOC1) + IOVA instrumentation. Full write-up in #49691.

## Fix
Only take the pinned fast-path when the buffer's IOVA range lies within the NOC-addressable PCIe window (`hal.get_pcie_addr_lower_bound()..get_pcie_addr_upper_bound()` — the same bounds the firmware uses for `noc_pcie_addr`); otherwise fall back to the non-pinned chunked hugepage path (always-addressable command-queue IOVA). The guard is on **IOVA addressability, not transfer size**.

## Regression test
`distributed_unit_tests` → `MeshBufferTestSuite.EnqueueWritePinnedMemoryLargeInterleavedBuffer`: a 640 MiB pinned interleaved DRAM write + read-back (large enough to reliably land a top-down IOVA on an IOMMU runner). **Verified: hangs without the guard, passes with it.**

## CI note
The pinned path only activates on **IOMMU (viommu)** runners. The gtest name matches the `*PinnedMemory*` filter used by the `metal-iommu-unit-tests` job (viommu-stable runners) — but that workflow isn't currently wired into an active pipeline. Follow-up: wire it in so this has standing CI coverage.

## Root fix (separate)
Constrain the device DMA mask so the IOMMU never allocates unreachable IOVAs — **#49696**. With that, large pinned writes work directly and this guard becomes a redundant safety net.

## Follow-ups
- Once merged, the `_skip_multi_chunk_hangs` skip on `pjosipovic/topk-large-indices-realtime-perf` can be removed.

Closes #49691

### CODE REVIEW

**Comment by pavlejosipovic** [comment]
Consolidated into #49697, which contains the primary NOC-DMA fix plus this addressability guard (as the old-KMD fallback) and the shared regression gtest. Closing in favor of #49697.
